app/src/main/java/org/xbmc/kore/jsonrpc/HostConnection.java

```diff
@@ -380,6 +409,51 @@ public void onError(int errorCode, String description) {
         }, null);
         return future;
     }
+    
+    /**
+     * Executes the {@link Callable} and waits for it to finish on a background thread. The
+     * result is returned using the {@link ApiCallback} and handler
+     * @param callable executed using an {@link ExecutorService}
+     * @param apiCallback used to return the result of the callable
+     * @param handler used to execute the {@link ApiCallback} methods
+     * @param <T>
+     */
+    public <T> void execute(Callable<T> callable, final ApiCallback<T> apiCallback, final Handler handler) {
```

Comment:
I think that this method isn't consistent with the other `execute`, or even with `HostConnection`. `execute` is defined as executing a jsonrpc method on the server, and return using either a `Future` or a callback.
This method is very generic - it runs a `callable` on a background thread, and posts the result on a callback (which is `Apicallback` just because it's handy). That has nothing to do with `HostConnection`, it's generic (and i don't like using `ApiCallback` here, because of that same reason).
Maybe we should think this more structurally. What is needed here? A way to execute several jsonrpc methods in sequence? I'm not convinced that we need to have something specific to do this (versus just creating a runnable and executing it in place in a thread), but if we need it it might make sense to create a specific class for it (a la `HostAction`, though i'm not convinced that it should be needed) or just look at `CompletableFuture`.

Anyway, this can be merged as is for now, but i plan to revisit this.

Reply:
Indeed it's more structural. I really don't like the threading logic in the activities and fragments. I think the activities and fragments should only manage the view elements and all else should be placed in other classes. I'm a fan of Clean Architecture and MVVM and if I had the time I would refactor Kore as a whole. But for the time being I would appreciate it if we could move as much logic away from the activities and fragments as possible. If I get to it I will propose a setup and plan to gradually refactor Kore. It will probably be something based on Clean Architecture, and MVVM using Kotlin with Coroutines and RXjava.

Reply:
I don't share your dislike of using threads in activities/fragments, as long as it's for a UI reason. Typically the reason to launch background threads is to free the UI thread so that it continues to be responsive, and that's intrinsically an UI issue, so i don't see why it shouldn't be where the UI is managed.
Anyway,  my issue with this method isn't related to that, it's just that i don't think its semantics are consistend with the rest of this class.